GCC 11 warns of UB in matrix_la.h

Noticed this warning when building a downstream, but this happens on current HEAD (cf21f5aa9556f31190b5668b8b159185f45361f0) too.

GCC 11.2.0-10 on sid x32, default config, notes:
```
[ 45%] Building CXX object dlib/CMakeFiles/dlib.dir/filtering/kalman_filter.cpp.o
In file included from /tmp/dlib/dlib/filtering/../matrix.h:11,
                 from /tmp/dlib/dlib/filtering/kalman_filter.h:7,
                 from /tmp/dlib/dlib/filtering/kalman_filter.cpp:6:
/tmp/dlib/dlib/filtering/../matrix/matrix_la.h: In function ‘long int dlib::svd4(dlib::svd_u_mode, bool, const dlib::matrix_exp<EXP>&, dlib::matrix<typename EXP::type, uM, uN, MM1, L1>&, dlib::matrix<typename EXP::type, qN, qX, MM2, L1>&, dlib::matrix<typename EXP::type, vM, vN, MM3, L1>&) [with EXP = dlib::matrix_op<dlib::op_trans<dlib::matrix_op<dlib::op_trans<dlib::matrix<double, 1, 2, dlib::memory_manager_stateless_kernel_1<char>, dlib::row_major_layout> > > > >; long int qN = 1; long int qX = 1; long int uM = 1; long int uN = 1; long int vM = 2; long int vN = 1; MM1 = dlib::memory_manager_stateless_kernel_1<char>; MM2 = dlib::memory_manager_stateless_kernel_1<char>; MM3 = dlib::memory_manager_stateless_kernel_1<char>; L1 = dlib::row_major_layout]’:
/tmp/dlib/dlib/filtering/../matrix/matrix_la.h:225:32: warning: iteration 1 invokes undefined behavior [-Waggressive-loop-optimizations]
  225 |             y = abs(q(i)) + abs(e(i));
      |                             ~~~^~~~~~
/tmp/dlib/dlib/filtering/../matrix/matrix_la.h:163:20: note: within this loop
  163 |         for (i=0; i<n; i++)
      |                   ~^~
```

--- a/dlib/matrix/matrix_la.h
+++ b/dlib/matrix/matrix_la.h
@@ -160,9 +160,10 @@ namespace dlib
 
         /* Householder's reduction to bidiagonal form. */
         g = x = 0.0;    
-        for (i=0; i<n; i++) 
+        auto ei = e.begin();
+        for (i=0; i<n; i++, ei++)
         {
-            e(i) = g;
+            *ei = g;
             s = 0.0;
             l = i + 1;
 
@@ -222,7 +223,7 @@ namespace dlib
                 } /* end j */
             } /* end s */
 
-            y = abs(q(i)) + abs(e(i));                         
+            y = abs(q(i)) + abs(*ei);
             if (y > x)
                 x = y;
         } /* end i */

